boot/zephyr/boot_serial_extensions.c

```diff
@@ -1,161 +1,30 @@
-    struct flash_area const *fap;
-    struct image_header hdr;
-    int rc;
-    int img_install_stat;
-
-    rc = BOOT_HOOK_CALL(boot_img_install_stat_hook, BOOT_HOOK_REGULAR,
-                        image_index, slot, &img_install_stat);
-    if (rc == BOOT_HOOK_REGULAR)
-    {
-        img_install_stat = 0;
-    }
-
-    rc = BOOT_HOOK_CALL(boot_read_image_header_hook, BOOT_HOOK_REGULAR,
-                        image_index, slot, &hdr);
-    if (rc == BOOT_HOOK_REGULAR)
-    {
-        area_id = flash_area_id_from_multi_image_slot(image_index, slot);
-
-        rc = flash_area_open(area_id, &fap);
-        if (rc) {
-            return rc;
-        }
-
-        rc = flash_area_read(fap, 0, &hdr, sizeof(hdr));
-
-        flash_area_close(fap);
-    }
-
-    if (rc == 0) {
-        if (hdr.ih_magic == IMAGE_MAGIC) {
-            snprintf(buffer, len, "ver=%d.%d.%d.%d,install_stat=%d",
-                    hdr.ih_ver.iv_major,
-                    hdr.ih_ver.iv_minor,
-                    hdr.ih_ver.iv_revision,
-                    hdr.ih_ver.iv_build_num,
-                    img_install_stat);
-        } else {
-            rc = 1;
-        }
-    }
-
-    return rc;
-}
-
-static int bs_custom_img_list(zcbor_state_t *cs)
-{
-    int rc = 0;
-    char tmpbuf[64];	/* Buffer should fit version and flags */
-
-    zcbor_map_start_encode(cs, 10);
-
-    for (int img = 0; img < MCUBOOT_IMAGE_NUMBER; img++) {
-        for (int slot = 0; slot < 2; slot++) {
-            rc = custom_img_status(img, slot, tmpbuf, sizeof(tmpbuf));
-
-            zcbor_int32_put(cs, img * 2 + slot + 1);
-            if (rc == 0) {
-                zcbor_tstr_put_term(cs, tmpbuf);
-            } else {
-                zcbor_tstr_put_lit(cs, "");
-            }
-        }
-    }
-
-    zcbor_tstr_put_lit(cs, "rc");
-    zcbor_uint32_put(cs, MGMT_ERR_OK);
-    zcbor_map_end_encode(cs, 10);
-
-    return rc;
-}
-
-#ifndef ZEPHYR_MGMT_GRP_BASIC_CMD_IMAGE_LIST
-    #define ZEPHYR_MGMT_GRP_BASIC_CMD_IMAGE_LIST 1
-#endif
-#endif /*MCUBOOT_MGMT_CUSTOM_IMG_LIST*/
-
 int bs_peruser_system_specific(const struct nmgr_hdr *hdr, const char *buffer,
                                int len, zcbor_state_t *cs)
 {
     int mgmt_rc = MGMT_ERR_ENOTSUP;
 
-    if (hdr->nh_group == ZEPHYR_MGMT_GRP_BASIC) {
-        if (hdr->nh_op == NMGR_OP_WRITE) {
-#ifdef CONFIG_BOOT_MGMT_CUSTOM_STORAGE_ERASE
-            if (hdr->nh_id == ZEPHYR_MGMT_GRP_BASIC_CMD_ERASE_STORAGE) {
-                mgmt_rc = bs_custom_storage_erase(cs);
-            }
-#endif
-        } else if (hdr->nh_op == NMGR_OP_READ) {
-#ifdef MCUBOOT_MGMT_CUSTOM_IMG_LIST
-            if (hdr->nh_id == ZEPHYR_MGMT_GRP_BASIC_CMD_IMAGE_LIST) {
-                mgmt_rc = bs_custom_img_list(cs);
+    STRUCT_SECTION_FOREACH(mcuboot_bs_custom_handlers, function) {
+        if (function->handler) {
+            mgmt_rc = function->handler(hdr, buffer, len, cs);
+
+            if (mgmt_rc != MGMT_ERR_ENOTSUP) {
+                break;
```

Comment:
Wouldn't this mean that there might be more than one handler called for the same buffer and mutiple responses being sent for one?

Reply:
No, it would be a 1:1 mapping, if there are 3 handlers registered, the first one will be called, if it is not valid for that then it must return MGMT_ERR_ENOTSUP, it will then check the next handler, if the next handler is valid, it will return something else (i.e. usually EOK but could return EINVAL or anything) and then the loop here will be broken out, the final handler would not be called